### ISSUE TITLE
problem within KeywordPresent (as see in window's title or CPU)

### ISSUE THREAD

**Issue by alaingdl** [gnudatalanguage/gdl#1131]
I don't know how to modify KeywordPresent (in envt.hpp & envt.cpp) in such a way, when we pass an undefined keyword, it is considered as not present ! Eg : 

```
z=!null
window, 0, title=z
;(or, direct)
window, 1, title=!null
;(or, even more direct)
window, 2, title=a
```

In GDL now, the Window's title is not set to GDL 0, 1, 2 for all the three cases. (or you can remove the number)

Since this code _if (e->KeywordPresent())_ is widely used in the code, I am afraid for other un-expected side effects ...
In other terms, if the keyword is present but undefined (or !null), it is considered as defined ...

Other related problem in CPU, if we (wrongly) try to change a parameter using a !NULL velue for the keyword, GDL complains, but IDL says nothing and change nothing (eg `cpu, TPOOL_NTHREADS=!null`)

**Comment by GillesDuvert** [comment]
Same as #1051, no?
The use of KeywordPresent() just means that the keyword was there (i.e., it's name was written in the commandline). But its value may still be undefined.
All the functions in envt (and other of course) would benefit from some minimal explanation.
It's for sure unexpected side-effects will occur if KeywordPresent() is somehow modified!

**Comment by alaingdl** [comment]
now working on that. If I succeed to overpass the ID "token" issue the PR is on way.

**Comment by alaingdl** [comment]
solve in #1137  for WINDOW and CPU 

For sure many other places may benefit from this smarter way to test whether the keyword is well assigned or not. More to come for HELP (issue #1102), delayed because I have a trouble with verbosity in GetString()

It would be great to have some time dedicated to scan each entry of KeywordPresent() and to estimated whether it could benefit from this fine tuning.

**Comment by alaingdl** [comment]
Whouaou, the KeywordPresent() issue is wider than expected :sob: 
e.g. : 
```
; g and n are not defined, then not set in the sens of KeywordSet() 
IDL> a=randomu(seed, 10, gamma=g, normal=n)

GDL> a=randomu(seed, 10, gamma=g, normal=n)
% RANDOMU: Conflicting keywords.
% Execution halted at: $MAIN$          
```

And I have some other cases ... I need help !! (just 461 entries ...)

**Comment by GillesDuvert** [comment]
I've created #1132 because there is a fundamental difference between GDL interpreter and IDL's . Possibly due to a regression at some point, it would be interesting to check that the issue was there in the (very) old days. 
But the KWPresent() behaviour is correct:  in the supported cases of 'undefined' keywords (output variable, such as in `b=min(a,max=c)` where c is undefined at start but filled with max(a) value at end, we just `e->AssureGlobalKW(maxIx)` before populating the output value.

KeywordPresentAndDefined() will have its uses for input keywords indeed.

The above issue in randomu calls for a new issue report, let's close this one and open another.
It is very simple: the if then else in the C code does not handle correctly (i.e., as IDL would do) these cases where the passed values for conflicting KWs are undefined. Obviously KeywordPresentAndDefined() must be used in this case. 
It would be preferable to have a more general function to deal with conflicting attributes, at 'env' level instead of relying on everybody being super clever each time a new command is added to GDL, and I suspect this must be declared at 'libinit' level.

**Comment by GillesDuvert** [comment]
solved in #1173
